Docs/init integrations (#348)
it-at-m/refarch: docs/integrations/s3.md

# S3 Integration

> Integration for CRUD operations on a s3 storage. Also used for file handling in other integrations.

Reviewer: ```suggestion
Integration for CRUD operations on a S3 storage. Also used for file handling in other integrations.
```

## Modules

The S3 integration follows the [default naming conventions](./index.md#naming-conventions).

Beside the default integration it contains different client libraries for accessing the integration. The client
libraries are especially provided for usage in other integrations.

> - REST: The rest client uses the rest endpoints of the s3-rest-service to manage data in s3.

Reviewer: ```suggestion
- REST: The REST client uses the REST endpoints of the `s3-rest-service` module to manage data in S3.
```

> - Java: The Java client directly uses the in ports of the s3-core.

Reviewer: ```suggestion
- Java: The Java client directly uses the in ports of the `s3-core` module.
```

### Dependency graph

The following graph shows the relationships between the various modules and how they interact and rely on each other.

```mermaid
flowchart RL
    rest-service --> rest-starter --> rest-core --> core
    rest-starter --> starter --> core
    
    java-client-starter --> java-client --> client-core
    java-client --> core
    java-client-starter --> starter
    rest-client-starter --> rest-client --> client-core

>     rest-client -.REST.-> rest-service

Reviewer: I actually don't like that only one connection is labeled as REST. See screenshot. 
![grafik](https://github.com/user-attachments/assets/23a2551a-c6c1-4ff6-89f6-033625f694ac)

In my opinion we either document all communication protocols or none of them.

Author: All other integrations do only have communication to external Service, so it's a different situation.
But from my side i would be ok to add also external service to all other graphs.

```

## Usage

```xml

<dependencies>
    <!-- REST -->

>     <!-- requires refarch-s3-integration-rest-service -->

Reviewer: If the `refarch-s3-integration-rest-service` is required why not directly include it here?
Or do you mean besides those dependency the  associated service has to run?

Reviewer: Then I'd mention that the service has to be running as an application. Currently it sounds like it has to be included as dependency, but this is exactly not what we want.

    <dependency>
        <groupId>de.muenchen.refarch</groupId>
        <artifactId>refarch-s3-integration-rest-starter</artifactId>
        <version>...</version>
    </dependency>
    <!-- or Java -->
    <dependency>
        <groupId>de.muenchen.refarch</groupId>
        <artifactId>refarch-s3-integration-java-starter</artifactId>
        <version>...</version>
    </dependency>
</dependencies>
```

## Configuration

Following are the properties to configure the different modules. Some of them are custom defined and others are synonyms
for spring package properties.
Whether a property is an alias can be checked in the corresponding `application.yml` of each module.

### s3-integration-rest-service

| Property                                        | Description                       | Example          |
|-------------------------------------------------|-----------------------------------|------------------|

> | `refarch.s3.url`                                | Url of s3 endpoint to connect to. | `s3.example.com` |

Reviewer: ```suggestion
| `refarch.s3.url`                                | URL of S3 endpoint to connect to. | `s3.example.com` |
```

| `refarch.s3.bucket-name`                        | Name of the bucket to connect to. | `refarch-bucket` |
| `refarch.s3.access-key`                         | Access key to use for connection. |                  |
| `refarch.s3.secret-key`                         | Secret key to use for connection. |                  |

> | `refarch.s3.initial-connection-test` (optional) | Test connection to s3 at startup  | `true` (default) |

Reviewer: ```suggestion
| `refarch.s3.initial-connection-test` (optional) | Test connection to S3 at startup.  | `true` (default) |
```

|------------------------------------------------------------|-------------------------------------------------------|--------------------------|
| `refarch.s3.client.max-file-size` (optional)               | Single file limit for up- or downloading in byte.     | `10MB`                   |
| `refarch.s3.client.max-batch-size` (optional)              | Limit for up- or downloading a list of files in byte. | `100MB`                  |
| `refarch.s3.client.supported-file-extensions.*` (optional) | Map of allowed file extensions for up- and download.  | `pdf: "application/pdf"` |

### s3-integration-rest-client-starter

All properties of [s3-integration-java-client-starter](#s3-integration-rest-client-starter) and following:

| Property                                 | Description                                                                | Example                                       |
|------------------------------------------|----------------------------------------------------------------------------|-----------------------------------------------|

> | `refarch.s3.client.document-storage-url` | Url to the RefArch S3 integration service.                                 | `http://s3-integration-service:8080`          |

Reviewer: ```suggestion
| `refarch.s3.client.document-storage-url` | URL to the RefArch S3 integration service.                                 | `http://s3-integration-service:8080`          |
```

> | `refarch.s3.client.enable-security`      | Switch to enable or disable oAuth2 authentication against s3 service.      | `true`                                        |

Reviewer: ```suggestion
| `refarch.s3.client.enable-security`      | Switch to enable or disable OAuth 2.0 authentication against S3 service.      | `true`                                        |
```

> For authentication against the s3-service a OAuth2 registration with the name `s3` needs to be provided.

Reviewer: ```suggestion
For authentication against the `s3-service` a OAuth 2.0 registration with the name `s3` needs to be provided.
```
